File: leetcode/middle/mst.16.19/mst16.19.cpp
class Solution 
{
public:

    inline void dfs(vector<vector<int>>& land, vector<int>& vis, int i, int j, int& cnt)
    {
        int n = land.size();
        int m = land[0].size();
        if(i<0 || i>=n || j<0 || j>=m || land[i][j] != 0 || vis[i*m+j] == 1)
            return;
        vis[i*m+j] = 1;
        cnt++;
        dfs(land, vis, i+1, j, cnt);
        dfs(land, vis, i-1, j, cnt);
        dfs(land, vis, i, j+1, cnt);
        dfs(land, vis, i, j-1, cnt);
        dfs(land, vis, i+1, j+1, cnt);
        dfs(land, vis, i+1, j-1, cnt);
        dfs(land, vis, i-1, j+1, cnt);
        dfs(land, vis, i-1, j-1, cnt);
    }
    vector<int> pondSizes(vector<vector<int>>& land) 
    {
        int n = land.size();
        int m = land[0].size();
        vector<vector<int>> vis(n, vector<int>(m, 0));
        vector<int> ans;
        for(int i = 0;i<n;i++)
        {
            for(int j = 0;j<m;j++)
            {
                if(land[i][j] == 0 && vis[i*m+j] == 0)
                {
                    int cnt = 0;
                    dfs(land, vis, i, j, cnt);
                    ans.push_back(cnt);
                }
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

File: leetcode/middle/mst.16.19/mst16.19.py
from typing import List
class Solution:
    def pondSizes(self, land: List[List[int]]) -> List[int]:
        ans : List[int] = []
        n = len(land)
        m = len(land[0])
        def dfs(i,j):
            if(i<0 or i>=n or j<0 or j>=m or land[i][j]!=0):
                return 0
            land[i][j] =-1
            lim = 1
            lim +=dfs(i+1,j)
            lim +=dfs(i-1,j)
            lim +=dfs(i,j+1)
            lim +=dfs(i,j-1)
            lim +=dfs(i+1,j+1)
            lim +=dfs(i+1,j-1)
            lim +=dfs(i-1,j-1)
            lim +=dfs(i-1,j+1)
            return lim
        for i in range(n):
            for j in range(m):
                if land[i][j] == 0:
                    ans.append(dfs(i,j))
        return sorted(ans)
